xapian-core/api/registry.cc

```diff
@@ -260,14 +322,21 @@ void
 Registry::register_weighting_scheme(const Xapian::Weight &wt)
 {
     LOGCALL_VOID(API, "Xapian::Registry::register_weighting_scheme", wt.name());
-    register_object(internal->wtschemes, wt);
+    register_object(internal->wtschemes, internal->wtschemes_short, wt);
 }
 
 const Xapian::Weight *
 Registry::get_weighting_scheme(const string & name) const
 {
     LOGCALL(API, const Xapian::Weight *, "Xapian::Registry::get_weighting_scheme", name);
-    RETURN(lookup_object(internal->wtschemes, name));
+    const Xapian::Weight * wt = lookup_object(internal->wtschemes, name);
+    const Xapian::Weight * wt_short = lookup_object(internal->wtschemes_short, name);
+    if (!name.empty() && C_islower(name[0])) {
+	if (wt_short == NULL)
+	    RETURN(wt);
+	RETURN(wt_short);
+    }
+    RETURN(wt);
```

Comment:
Here you look up both `wt` and `wt_short` every time, which is wasted effort in most cases.

You only need to look up `wt_short` if the lower case initial check succeeds.

And you only need to look up `wt` after you're done with potentially checking the short name - if you found an object by the short name, you can just return it.